cloud/services/compute/loadbalancers/reconcile.go

```diff
@@ -18,72 +18,228 @@ package loadbalancers
+		}
+		if err := s.deleteInternalLoadBalancer(ctx, name); err != nil {
+			return err
+		}
+	}
+
+	// Delete shared instancegroups last
+	return s.deleteInstanceGroups(ctx)
+}
+
+func (s *Service) deleteExternalLoadBalancer(ctx context.Context) error {
+	log := log.FromContext(ctx)
+	log.Info("Deleting external loadbalancer resources")
+	name := infrav1.APIServerRoleTagValue
+	if err := s.deleteForwardingRule(ctx, name); err != nil {
+		return err
+	}
+	s.scope.Network().APIServerForwardingRule = nil
+
+	if err := s.deleteAddress(ctx, name); err != nil {
+		return err
+	}
+	s.scope.Network().APIServerAddress = nil
+
+	if err := s.deleteTargetTCPProxy(ctx); err != nil {
+		return err
+	}
+	s.scope.Network().APIServerTargetProxy = nil
+
+	if err := s.deleteBackendService(ctx, name); err != nil {
+		return err
+	}
+	s.scope.Network().APIServerBackendService = nil
+
+	if err := s.deleteHealthCheck(ctx, name); err != nil {
+		return err
+	}
+	s.scope.Network().APIServerHealthCheck = nil
+
+	return nil
+}
+
+func (s *Service) deleteInternalLoadBalancer(ctx context.Context, name string) error {
+	log := log.FromContext(ctx)
+	log.Info("Deleting internal loadbalancer resources")
+	if err := s.deleteForwardingRule(ctx, name); err != nil {
+		return err
+	}
+	s.scope.Network().APIInternalForwardingRule = nil
+
+	if err := s.deleteAddress(ctx, name); err != nil {
+		return err
+	}
+	s.scope.Network().APIInternalAddress = nil
+
+	if err := s.deleteBackendService(ctx, name); err != nil {
+		return err
+	}
+	s.scope.Network().APIInternalBackendService = nil
+
+	if err := s.deleteHealthCheck(ctx, name); err != nil {
+		return err
+	}
+	s.scope.Network().APIInternalHealthCheck = nil
+
+	return nil
+}
+
+// createExternalLoadBalancer creates the components for a Global External Proxy LoadBalancer.
+func (s *Service) createExternalLoadBalancer(ctx context.Context, lbType infrav1.LoadBalancerType, instancegroups []*compute.InstanceGroup) error {
+	name := infrav1.APIServerRoleTagValue
+	healthcheck, err := s.createOrGetHealthCheck(ctx, name)
 	if err != nil {
 		return err
 	}
+	s.scope.Network().APIServerHealthCheck = ptr.To[string](healthcheck.SelfLink)
 
-	backendsvc, err := s.createOrGetBackendService(ctx, instancegroups, healthcheck)
+	// If an Internal LoadBalancer is being created, the BalancingMode must match the Internal LB
+	mode := loadBalancingModeUtilization
+	if lbType == infrav1.DualLoadBalancer {
+		mode = loadBalancingModeConnection
+	}
+	backendsvc, err := s.createOrGetBackendService(ctx, name, mode, instancegroups, healthcheck)
 	if err != nil {
 		return err
 	}
+	s.scope.Network().APIServerBackendService = ptr.To[string](backendsvc.SelfLink)
 
+	// Create TargetTCPProxy for Proxy Load Balancer
 	target, err := s.createOrGetTargetTCPProxy(ctx, backendsvc)
 	if err != nil {
 		return err
 	}
+	s.scope.Network().APIServerTargetProxy = ptr.To[string](target.SelfLink)
```

Comment:
Is the target proxy not used at all for the internal load balancer? I could see where we only need one in the case where the external LB is being used as an internal passthrough, but I'm unclear on whether it's required for internal-only.

I'm not sure which line in [this chart](https://cloud.google.com/load-balancing/docs/target-proxies#target_proxy_types) is being implemented, or if that chart even applies.

Reply:
Correct, the target TCP proxy is not required for an internal load balancer which is passthrough only, see https://cloud.google.com/kubernetes-engine/docs/concepts/service-load-balancer
`Internal LoadBalancer Services are implemented by using internal passthrough Network Load Balancers.`

Also https://cloud.google.com/load-balancing/docs/internal